### PR TITLE
PF-1442: start ssh agent and clone all the git repos references

### PR DESCRIPTION
**Pull request by yuhuyoyo**

### CODE REVIEW

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh [hunk 1]
```diff
@@ -76,9 +76,12 @@ fi
 # Log in with app-default-credentials
 sudo -u "${JUPYTER_USER}" sh -c "terra auth login --mode=APP_DEFAULT_CREDENTIALS"
 
-# Get all the git repvoio clone urls from the workspace resource list with type GIT_REPO and try to clone all of them.
+sudo -u "${JUPYTER_USER}" sh -c "mkdir -p /home/${JUPYTER_USER}/.ssh"
+cd /home/${JUPYTER_USER}/.ssh
+sudo -u "${JUPYTER_USER}" sh -c "terra user ssh-key get --format=JSON | jq -r '.privateSshKey'> id_rsa"
```
Comment (zloery) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh:81]: What happens if the user hasn't already registered an SSH key? I don't know how notebooks handle failure in startup scripts. 

also, the previous approach allowed users to clone public repos without setting up an ssh key with Github, but it seems like that wouldn't happen now. Is that behavior we care about, or was that just useful before we had ssh keys?

Comment (cbookg) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh:81]: +1, I think we still need to try the repos below even if the key fails.  and likely ensure we try them all such that we can clone a public one even if a private one fails?

Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh]: good point. done

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh [hunk 2]
```diff
@@ -76,9 +76,15 @@ fi
...
+sudo -u "${JUPYTER_USER}" sh -c "mkdir -p /home/${JUPYTER_USER}/.ssh"
 cd /home/${JUPYTER_USER}
-sudo -u "${JUPYTER_USER}" sh -c "terra resource list --type=GIT_REPO --format=JSON" | jq -r 'unique_by(.gitRepoUrl) | .[] | {gitRepoUrl} | join (" ") ' | while read -r repoUrl; do
-  sudo -u "${JUPYTER_USER}" sh -c "git clone ${repoUrl}" && echo "cloned ${repoUrl}" || echo "clone ${repoUrl} failed"
-done
+readonly TERRA_SSH_KEY=$(sudo -u "${JUPYTER_USER}" sh -c "terra user ssh-key get --format=JSON")
+if [[ -n "${TERRA_SSH_KEY}" ]]; then
+  printf '%s' "${TERRA_SSH_KEY}" | sudo -u "${JUPYTER_USER}" sh -c "jq -r '.privateSshKey'> .ssh/id_rsa"
+  sudo -u "${JUPYTER_USER}" sh -c 'chmod go-rwx .ssh/id_rsa'
+  echo eval '"$(ssh-agent -s)"' >> .bash_profile
```
Comment (cbookg) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh:85]: is this starting the agent?  i think we should start the agent even if we don't have a key

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh [hunk 3]
```diff
@@ -76,9 +76,15 @@ fi
...
 
-# Get all the git repvoio clone urls from the workspace resource list with type GIT_REPO and try to clone all of them.
+sudo -u "${JUPYTER_USER}" sh -c "mkdir -p /home/${JUPYTER_USER}/.ssh"
 cd /home/${JUPYTER_USER}
-sudo -u "${JUPYTER_USER}" sh -c "terra resource list --type=GIT_REPO --format=JSON" | jq -r 'unique_by(.gitRepoUrl) | .[] | {gitRepoUrl} | join (" ") ' | while read -r repoUrl; do
-  sudo -u "${JUPYTER_USER}" sh -c "git clone ${repoUrl}" && echo "cloned ${repoUrl}" || echo "clone ${repoUrl} failed"
-done
+readonly TERRA_SSH_KEY=$(sudo -u "${JUPYTER_USER}" sh -c "terra user ssh-key get --format=JSON")
+if [[ -n "${TERRA_SSH_KEY}" ]]; then
+  printf '%s' "${TERRA_SSH_KEY}" | sudo -u "${JUPYTER_USER}" sh -c "jq -r '.privateSshKey'> .ssh/id_rsa"
```
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh:83]: nit: space before `>`.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh [hunk 4]
```diff
@@ -76,9 +76,15 @@ fi
...
+  sudo -u "${JUPYTER_USER}" sh -c 'chmod go-rwx .ssh/id_rsa'
+  echo eval '"$(ssh-agent -s)"' >> .bash_profile
+  sudo -u "${JUPYTER_USER}" sh -c 'eval `ssh-agent -s`; ssh-add .ssh/id_rsa; ssh-keyscan -H github.com >> ~/.ssh/known_hosts'
+fi
+
+sudo -u "${JUPYTER_USER}" sh -c 'terra git clone --all'
```
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh:89]: So right now we're limited to one key for all repos (except public ones). I'm trying to think if that impacts the script in any way or if we should just document that.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh [hunk 5]
```diff
@@ -76,9 +76,15 @@ fi
 # Log in with app-default-credentials
 sudo -u "${JUPYTER_USER}" sh -c "terra auth login --mode=APP_DEFAULT_CREDENTIALS"
```
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh:77]: I don't _think_ we need the curly braces here. Does `"$JUPYPTER_USER"` amount to the same thing?

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh [hunk 6]
```diff
@@ -76,9 +76,15 @@ fi
...
 
-# Get all the git repvoio clone urls from the workspace resource list with type GIT_REPO and try to clone all of them.
+sudo -u "${JUPYTER_USER}" sh -c "mkdir -p /home/${JUPYTER_USER}/.ssh"
 cd /home/${JUPYTER_USER}
-sudo -u "${JUPYTER_USER}" sh -c "terra resource list --type=GIT_REPO --format=JSON" | jq -r 'unique_by(.gitRepoUrl) | .[] | {gitRepoUrl} | join (" ") ' | while read -r repoUrl; do
-  sudo -u "${JUPYTER_USER}" sh -c "git clone ${repoUrl}" && echo "cloned ${repoUrl}" || echo "clone ${repoUrl} failed"
-done
+readonly TERRA_SSH_KEY=$(sudo -u "${JUPYTER_USER}" sh -c "terra user ssh-key get --format=JSON")
```
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh:81]: Does this error out if there's no SSH key, or just return an empty structure?

**Review by jaycarlton — APPROVED** [review]

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh (hunk 5 above)
Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh:77]: i believe it does. i just wanted to keep the style consistent. i can remove it here tho..

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh (hunk 6 above)
Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh:81]: it returns empty

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh (hunk 2 above)
Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh]: done.

**Review by zloery — APPROVED** [review]

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh [hunk 7]
```diff
@@ -76,9 +76,19 @@ fi
...
+
+# Start the ssh-agent. Set this command in bash_profile so everytime user starts a shell, we start the ssh-agent.
+echo eval '"$(ssh-agent -s)"' >> .bash_profile
+if [[ -n "$TERRA_SSH_KEY" ]]; then
+  printf '%s' "$TERRA_SSH_KEY" | sudo -u "${JUPYTER_USER}" sh -c "jq -r '.privateSshKey' > .ssh/id_rsa"
+  sudo -u "$JUPYTER_USER" sh -c 'chmod go-rwx .ssh/id_rsa'
+  sudo -u "$JUPYTER_USER" sh -c 'eval `ssh-agent -s`; ssh-add .ssh/id_rsa; ssh-keyscan -H github.com >> ~/.ssh/known_hosts'
```
Comment (zloery) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh:88]: Line 84 has `echo eval '"$(ssh-agent -s)"' >> .bash_profile` is already a line above, do you also need `eval `ssh-agent -s` here?

Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/post-startup.sh]: good catch. this was left over from before i added line 84.

**Pull request merged**